src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp

```diff
@@ -0,0 +1,299 @@
+            setup_point, setup_angle, 0.0, TbotsProto::DribblerMode::OFF,
+            TbotsProto::BallCollisionType::AVOID, {AutoChipOrKickMode::OFF, 0},
+            TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT, 0.0);
+
+        tactics_to_run[0].emplace_back(align_placement_tactic);
+
+        event.common.set_tactics(tactics_to_run);
+    }
+}
+
+void BallPlacementPlayFSM::placeBall(const Update &event)
+{
+    PriorityTacticVector tactics_to_run = {{}};
+
+    // setup move tactics for robots away from ball placing robot
+    setupMoveTactics(event);
+    tactics_to_run[0].insert(tactics_to_run[0].end(), move_tactics.begin(),
+                             move_tactics.end());
+
+    Angle final_angle = Angle::zero();
+    std::optional<Point> placement_point =
+        event.common.world.gameState().getBallPlacementPoint();
+
+    Vector placement_dribble_vector;
+    if (placement_point.has_value())
+    {
+        placement_dribble_vector =
+            placement_point.value() - event.common.world.ball().position();
+        final_angle = placement_dribble_vector.orientation();
+    }
+
+    // setup ball placement tactic for ball placing robot
+    place_ball_tactic->updateControlParams(
+        event.common.world.gameState().getBallPlacementPoint(), final_angle, true);
+    tactics_to_run[0].emplace_back(place_ball_tactic);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+void BallPlacementPlayFSM::startWait(const Update &event)
+{
+    start_time = std::chrono::system_clock::now();
+}
+
+void BallPlacementPlayFSM::retreat(const Update &event)
+{
+    World world = event.common.world;
+    std::optional<Robot> nearest_robot =
+        world.friendlyTeam().getNearestRobot(world.ball().position());
+
+    if (nearest_robot.has_value())
+    {
+        PriorityTacticVector tactics_to_run = {{}};
+
+        // setup move tactics for robots away from ball placing robot
+        setupMoveTactics(event);
+        tactics_to_run[0].insert(tactics_to_run[0].end(), move_tactics.begin(),
+                                 move_tactics.end());
+
+        Point ball_pos = world.ball().position();
+
+        // robot will try to retreat backwards from wherever it is currently facing
+        Angle final_orientation  = nearest_robot.value().orientation();
+        Vector retreat_direction = (nearest_robot->position() - ball_pos).normalize();
+        Point retreat_position =
+            ball_pos + retreat_direction * (0.5 + ROBOT_MAX_RADIUS_METERS);
+
+        // if the initial retreat position is out of the field boundary, have it retreat
+        // towards the closest goal
+        if (!contains(world.field().fieldBoundary(), retreat_position))
+        {
+            bool in_friendly_half = contains(world.field().friendlyHalf(), ball_pos);
+            Point closer_goal     = in_friendly_half ? world.field().friendlyGoalCenter()
+                                                 : world.field().enemyGoalCenter();
```

Comment:
as part of our codestyle guidelines, we prefer:
```suggestion
            Point closer_goal     = world.friendlyGoalCenter()
            if (!in_friendly_half)
            {
                world.field().enemyGoalCenter();
            }
```